Core/include/Acts/Utilities/ProtoAxis.hpp

```diff
@@ -0,0 +1,134 @@
+            double maxE, std::size_t nbins);
+
+  /// Placeholder constructor for auto-range binning
+  ///
+  /// @param aDir the value/cast in which this is binned
+  /// @param abType the axis boundary type
+  /// @param nbins the number of bins
+  ///
+  /// @note that auto-range is only supported for equidistant binning
+  ProtoAxis(AxisDirection aDir, AxisBoundaryType abType, std::size_t nbins);
+
+  ProtoAxis(const ProtoAxis&) = delete;
+  ProtoAxis& operator=(const ProtoAxis&) = delete;
+  ProtoAxis(ProtoAxis&&) = default;
+  ProtoAxis& operator=(ProtoAxis&&) = default;
+
+  /// @brief returns the axis direction
+  ///
+  /// @return @c AxisDirection of this axis
+  AxisDirection getAxisDirection() const;
+
+  /// @brief Return the IAxis representation
+  ///
+  /// @return @c AxisType of this axis
+  const IAxis& getAxis() const;
+
+  /// @brief check if this is an auto-range binning
+  bool isAutorange() const;
+
+  /// Dump into a string
+  /// @return the string representation
+  std::string toString() const;
+
+ protected:
+  /// Dispatch to the correct stream operator
+  /// @param os output stream
+  void toStream(std::ostream& os) const;
+
+  /// The axis direction
+  AxisDirection m_axisDir = AxisDirection::AxisX;
+
+  /// The axis representation
+  std::unique_ptr<IAxis> m_axis = nullptr;
+
+  /// Indicate if this is a place holder auto-range binning
+  bool m_autorange = false;
+};
+
+/// @brief Helper method to create a 1D grid from a single proto axis
+///
+/// @tparam payload_t the grid payloat type
+///
+/// @param a the proto axis
+///
+/// @return an IGrid unique ptr and hence transfers ownership
+template <typename payload_t>
+std::unique_ptr<IGrid> makeGrid(const ProtoAxis& a) {
+  return a.getAxis().visit([&](const auto& axis) -> std::unique_ptr<IGrid> {
+    using AxisTypeA = std::decay_t<decltype(axis)>;
+    using GridType = Grid<payload_t, AxisTypeA>;
+    return std::make_unique<GridType>(axis);
+  });
+}
+
+/// @brief Helper method to create a 2D grid from a two proto axes
+///
+/// @tparam payload_t the grid payloat type
+///
+/// @param a the first proto axis
+/// @param b the second proto axis
+///
+/// @return an IGrid unique ptr and hence transfers ownership
+template <typename payload_t>
+std::unique_ptr<IGrid> makeGrid(const ProtoAxis& a, const ProtoAxis& b) {
+  // Validate axis compatibility
+  if (a.getAxisDirection() == b.getAxisDirection()) {
+    throw std::invalid_argument(
+        "ProtoAxis::makeGrid: Axes must have different directions");
+  }
+  return a.getAxis().visit([&](const auto& axisA) -> std::unique_ptr<IGrid> {
+    return b.getAxis().visit([&](const auto& axisB) -> std::unique_ptr<IGrid> {
+      using AxisTypeA = std::decay_t<decltype(axisA)>;
+      using AxisTypeB = std::decay_t<decltype(axisB)>;
+      using GridType = Grid<payload_t, AxisTypeA, AxisTypeB>;
+      return std::make_unique<GridType>(axisA, axisB);
+    });
+  });
+}
```

Comment:
Not a must, just a comment: since C++20 its possible to have explicit template parameters on lambdas, like:

```suggestion
  return a.getAxis().visit([&]<typename AxisTypeA>(const AxisTypeA& axisA) -> std::unique_ptr<IGrid> {
    return b.getAxis().visit([&]<typename AxisTypeB>(const AxisTypeB& axisB) -> std::unique_ptr<IGrid> {
      using GridType = Grid<payload_t, AxisTypeA, AxisTypeB>;
      return std::make_unique<GridType>(axisA, axisB);
    });
  });
}
```

which in this case saves you from a bit of gymnastics.

Reply:
Cool, I tried that, but didn't really get it to work ... will accept.